Dockerfile: switch from CGI to FastCGI for PHP

schildbach opened the pull request:
Disclaimer:

This PR requires extra review scrutiny, as I've got no experience with PHP. I'm already running this PR plus an HTTP/2 (see #8349) enablement on my private server and will look into any problems that might arise. So far, it feels very performant.

Changes proposed in this pull request:

Switch from CGI to FastCGI for PHP when running inside Docker

How to test the feature manually:

1. build and run the Debian docker image
2. test UI via web browser
3. test API via reader apps (Newsflash etc)

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [ ] unit tests written (optional if too hard)
- [ ] documentation updated

schildbach committed:
Dockerfile: switch from CGI to FastCGI for PHP

This installs and runs PHP-FPM (FastCGI process manager).

Both Apache2 and PHP-FPM are put into background as a shell job,
such that "wait -n" can be used to terminate if either of both
exits.

If any of the jobs outputs to standard out/err, it will go to
the Docker log.

Frenzie commented on Docker/Dockerfile:
```diff
@@ -69,5 +71,6 @@ ENTRYPOINT ["./Docker/entrypoint.sh"]
 EXPOSE 80
 # hadolint ignore=DL3025
 CMD ([ -z "$CRON_MIN" ] || cron) && \
+	/usr/sbin/php-fpm8.4 -F & \
```
Is that a typo or meant to be a background process? It looks odd in this location either way.

schildbach replied:
Yes, it's meant to be run in the background. Otherwise, Apache (down below) would never be started.

schildbach replied:
The idea of the construct is that both Apache and FPM need to run in parallel, and both shall send their stdout/stderr to the console so it's sent to the Docker log.

Frenzie replied:
I'll attempt to clarify what I mean.

1. Compare the difference between these two commands:
   ```
   false && true
   false && { false; } & true
   ```
   The behavior is quite different. The former won't execute the second/third (Apache), while the latter will. This change in behavior won't occur if you put the background process first.
   ```
   { false; } & false && true
   ```

2. The entire `-F &`. Keep it in foreground, then put it in the background. That sounds a bit weird. How is it different from a plain `/usr/sbin/php-fpm8.4 -D;`? Something related to logging? Doesn't it start in the background by default regardless?
3. How dependable is it in case of errors? How much harder is it to debug? Shouldn't it be managed by supervisord or by separate Docker containers instead?

schildbach replied:
ad 1: So you suggest to put the FPM line before the cron line?

ad 2: Yes, the idea is that FPM sends its output to stdout/stderr so it appears in the Docker logs. Try `echo hello &` in a shell and you'll see the "hello". Deamonizing the process won't do this. But if you prefer we can use `-D` rather than the shell backgrounding (`&`).

ad 3: FPM needs to run, otherwise the app won't work (apart from some staticly served resources maybe). Debugging I can't tell, but I tried to preserve log output (see 2). For development the old configuration can stilll be used if desired. A separate container for FPM would probably be possible, though it's a more complicated setup that will require some kind of orchestrator (e.g. compose).

schildbach replied:
Did my last two patches address your points?

Frenzie replied:
> ad 1: So you suggest to put the FPM line before the cron line?

Unless the change in behavior is on purpose. I'm just saying it looks weird and suspicious to change the behavior. These things jump out. ;-)

> ad 2: Yes, the idea is that FPM sends its output to stdout/stderr so it appears in the Docker logs. Try echo hello & in a shell and you'll see the "hello". Deamonizing the process won't do this. But if you prefer we can use -D rather than the shell backgrounding (&).

It's less about any particular choice as it is about documenting why the choice was made. Even you might not know anymore a month from now and then you just have some mystery flags that require extensive research and testing to adjust.

> ad 3: FPM needs to run, otherwise the app won't work (apart from some staticly served resources maybe). 

That's my point. Something like supervisord would ensure it actually runs. Right now it doesn't matter because Docker will always make sure Apache runs (that is, the container will restart in the unlikely event Apache should crash), but it won't for this background process. Presumably and hopefully this is an almost entirely academic concern, but it seems to be a potential point of failure that doesn't currently exist.

schildbach replied:
I looked at supervisord but it seems quite involved. I'd rather not add so much complexity, if it's only academic.

Frenzie replied:
It's a very real risk. What I mean is that PHP is probably fairly battle tested by now.

schildbach replied:
Apparently we could use this pattern:

```
process1 &   # fpm
process2 &   # apache
wait -nf         # wait for any of the two processes to finish
```

However this seems to work only with shell backgrounding, not with deamonizing.

Frenzie replied:
That pattern makes a lot of sense to me (with the -F then).

schildbach replied:
Ok, I'll try that. Will take until first week of Jan though.

schildbach replied:
Done.

schildbach commented:
I just tried how a PHP error would look like and it's logged by the FastCGI proxy anyway:

```
[Fri Dec 26 12:35:44.066206 2025] [proxy_fcgi:error] [pid 38:tid 43] [client fd00::1:47838] AH01071: Got error 'PHP message: PHP Fatal error:  Uncaught Error: Undefined constant "blablablbala" in /var/www/FreshRSS/p/i/index.php:31\nStack trace:\n#0 {main}\n  thrown in /var/www/FreshRSS/p/i/index.php on line 31'
```

So I pushed a fix on top that changes the shell backgrounding to daemonizing (`-D` switch).

schildbach commented:
And another patch adds a symlink from the FPM error log to stderr. This causes FPM startup messages be visible in the Docker log, and I assume also serious errors would show up.

Alkarex commented:
Thanks for this PR @schildbach 
This is a nice way to start testing.

Apparently, `flush()` is not working https://github.com/FreshRSS/FreshRSS/issues/8349#issuecomment-3691649306
So that will require a bit more research

Alkarex commented on Docker/Dockerfile:
```diff
@@ -7,7 +7,7 @@ ARG DEBIAN_FRONTEND=noninteractive
 	apt-get install --no-install-recommends -y \
 	ca-certificates cron \
-	apache2 libapache2-mod-php \
+	apache2 libapache2-mod-fcgid php-fpm \
```
I am not sure of the exact pros and cons for the FreshRSS main use-cases, but https://cwiki.apache.org/confluence/display/HTTPD/Php-fcgid tells:

> On what conditions not to run php on fcgid
> If you run httpd 2.4, you should consider [PHP-FPM](http://wiki.apache.org/httpd/PHP-FPM)

And https://cwiki.apache.org/confluence/display/httpd/PHP-FPM tells:

> using mod_proxy_fcgi and php-fpm

schildbach replied:
Fwiw, the Docker image is using Apache/2.4.65 (Debian).

schildbach commented:
I just implemented the "process1 & ; process2 & ; wait -n" pattern described above, and successfully tested that it works as expected by killing either of the processes.

I also squashed and rebased to edge. I also added a more detailed commit description.

So I think what's left is the `flush()` issue. Can some PHP dev help with this? I have zero PHP experience. Where in the app does the issue surface?

Frenzie commented:
It can be a big speed difference if you set the network to emulate some slow speed (3G or less), especially if the server and/or DB is also being a bit slow.

Frenzie commented:
Basically it means output should be sent ASAP and not at some later point when the full page is done.

Edit: you can test really easily by inserting a sleep(2) or some such in a basic for loop.

Alkarex commented:
> So I think what's left is the `flush()` issue. Can some PHP dev help with this? I have zero PHP experience. Where in the app does the issue surface?

This is not specific to PHP. It is the same problem with any other language. The flush cannot be controlled from us anymore because there are suddenly several layers, each with their own buffer.

Explanation + how to reproduce: https://github.com/FreshRSS/FreshRSS/issues/8349#issuecomment-3691649306

The main issue is that users are faced with a white page for potentially a long time when the database request is slow, which is the case for instance when making a search. Before this PR, since we could control the buffering, a partial page is sent so that at least the menus and a waiting animation is shown.

Another issue is that time to first byte gets poorer, leading to reduced performance especially on slow networks.

I think the options include reducing the buffering size at some of the new layers introduced by this PR, but this will reduce the performances in some other situations. I am not sure there is an easy solution for that.

A first step in any case would be to add a performance test, which can be used to compare various approaches (not only for this PR). I have some elements of that on my own TODO list.

schildbach commented:
Ok, thanks, now I understand the problem. Let me ask two questions:

- How does the common case of a reverse proxy relate to the issue? Is `flush()` relayed e.g. via Traefik? My understanding is if it doesn't "arrive" at the user's web browser, `flush()` is ineffective.
- Isn't this kind of reliance on long-running requests a bad design anyway? I'd expect to receive a quick response with the page layout and a loading indicator (in place of the DB results), and then a second (background, long-running) request queries the DB?

Alkarex commented:
> How does the common case of a reverse proxy relate to the issue? Is `flush()` relayed e.g. via Traefik? My understanding is if it doesn't "arrive" at the user's web browser, `flush()` is ineffective.

I have not tested in depth, but I run with Traefik myself and it works fine. Additional investigation welcome.

> Isn't this kind of reliance on long-running requests a bad design anyway?

That requires a rewrite using JavaScript to load the data, which has pros and cons. I do plan to propose an alternative pure-JS UI at some point in the future, but I like the idea of having a true HTML response.

> I'd expect to receive a quick response with the page layout and a loading indicator (in place of the DB results), and then a second (background, long-running) request queries the DB?

This is the behaviour today, but with a single query.

Frenzie commented:
> Isn't this kind of reliance on long-running requests a bad design anyway? I'd expect to receive a quick response with the page layout and a loading indicator (in place of the DB results), and then a second (background, long-running) request queries the DB?

No, not in the slightest. The exact opposite is true. This is how you make fast, *dependable* websites that always work no matter what happens.
